data-access/Helpers/Exceptions.cs

```diff
@@ -1,38 +1,151 @@
 ﻿using System;
+using System.Net;
+using Newtonsoft.Json;
+using SQE.SqeHttpApi.DataAccess.Models;
 
 namespace SQE.SqeHttpApi.DataAccess.Helpers
 {
-    // TODO: Decide whether the controller is aware of these exceptions, or the service translates them
-    public abstract class RepositoryException: Exception
+    // Exceptions of this class will be caught by the middleware and thrown back to HTTP requests with the proper
+    // response codes.  This enables a RESTful experience regardless of whether a request is made via HTTP or by
+    // websocket.  In either case the HTTP router or the Hub will catch a child of the HttpException class
+    // and return that information to the user.
+    // Errors should be thrown properly at the lowest levels of code, this will reduce boiler-plate in higher
+    // level functions like the Services and ensure more consistent responses.
+    public class HttpException : Exception
```

Comment:
Is `HttpException` the right name? It is thrown by repositories, which are also used with SignalR.

Reply:
I'm open to another name.  You are right, these will get caught by signalr too and sent on to the user via websocket connections too.  I guess I thought of them as HttpExceptions since they map a more extensive internal error code system to the HTTP error codes. Do you prefer a different name?

Reply:
How about `ServerException` or `APIException`?

Reply:
I like APIException.  Also, I think you mentioned that each exception should be its own (sub)class, I liked that idea too.  It should not be too hard to implement that.

Reply:
Good. Just group them into groups of exceptions, so all permissions errors derive from `APIPermissionException`. It will allow you catch some of them in your middleware without writing numerous catch clauses (that are going to malfunction if you add another exception of that type)

Reply:
I do not have strong feelings towards taking the HTTP code out of these classes and limiting it to the middleware. It's obviously the more organized thing to do, but I think we can safely ignore this one. Putting the HTTP code inside the Exception implementation is probably a bit easier.

Reply:
Ok. They now follow what one would expect from common OOP practice.  I have an abstract base class, then abstract subclasses for the groups of exceptions, and then each individual exception class inherits from one of those.  As for the error codes (as mentioned below), I use the class name as the internal error code, which is simple enough and probably a nice feature for debugging and avoiding naming errors (you can't have two classes with the same name in the same namespace).